homeassistant/components/doorbird.py

```diff
@@ -155,13 +136,112 @@ def custom_url(self):
+                             sensor_type, self.name, DOMAIN, slug)
+            else:
+                # Disabled -> deregister
+                self._deregister_event(url, sensor_type)
+                _LOGGER.info('Deregistered %s pushes from DoorBird "%s". '
+                             'If any old favorites or schedules remain, '
+                             'follow the instructions in the component '
+                             'documentation to clear device registrations.',
+                             sensor_type, self.name)
+
+    def _register_event(self, hass_url, event, schedule):
+        """Add a schedule entry in the device for a sensor."""
+        from doorbirdpy import DoorBirdScheduleEntry,\
+            DoorBirdScheduleEntryOutput
+
+        # Register HA URL as favorite if not already, then get the ID
+        if not self.ha_is_favorite(hass_url):
+            self.device.change_favorite('http',
+                                        'Home Assistant on {} ({} events)'
+                                        .format(hass_url, event), hass_url)
+        fav_id = self.ha_favorite_id(hass_url)
+
+        # Add event handling to device schedule
+        entry = DoorBirdScheduleEntry(event)
+        entry.output.append(DoorBirdScheduleEntryOutput(event='http',
+                                                        param=fav_id,
+                                                        schedule=schedule))
+
+        if event == 'doorbell':
+            # Repeat edit for each monitored doorbell number
+            for doorbell in self.doorbell_nums:
+                entry.param = str(doorbell)
+                self.device.change_schedule(entry)
+        else:
+            self.device.change_schedule(entry)
 
-        return self._monitored_events
+    def _deregister_event(self, hass_url, event):
+        """Remove the schedule entry in the device for a sensor."""
+        # Find the right favorite and delete it
+        fav_id = self.ha_favorite_id(hass_url)
+        self._device.delete_favorite('http', fav_id)
 
+        # Delete the matching schedule for each doorbell number
+        for doorbell in self.doorbell_nums:
+            self._device.delete_schedule(event, str(doorbell))
 
-class DoorbirdRequestView(HomeAssistantView):
+    def ha_is_favorite(self, ha_url, favs=None) -> bool:
+        """Return whether the given URL is registered as a device favorite."""
+        favs = favs if favs else self.device.favorites()
+
+        if 'http' not in favs:
+            return False
+
+        for fav in favs['http'].values():
+            if fav['value'] == ha_url:
+                return True
+
+        return False
+
+    def ha_favorite_id(self, ha_url, favs=None) -> str:
+        """
+        Return the device favorite ID for the given URL.
+
+        The favorite must exist or there will be problems.
+        """
+        favs = favs if favs else self.device.favorites()
+
+        for fav_id in favs['http']:
```

Comment:
It is possible to delete all favorites via the Doorbird app which will cause a KeyError because 'http' key is missing.

Reply:
This is an old version of the code, this function is now called `get_ha_webhook_id`. This is not an issue because the favorite is added (creating an `http` object if one does not already exist) if `webhook_is_registered` (above: `ha_is_favorite`) returns false (and it handles a missing `http` object gracefully; see lines 222-223 above).

Reply:
Do you have changes that aren't pushed?  Ran into that while testing the schedule clearing.

Reply:
I thought I had pushed them (just renaming some things), but apparent I made them against the branch that didn't have your changes on it :man_facepalming: